dali/kernels/shape.h

```diff
@@ -0,0 +1,609 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SHAPE_H_
+#define DALI_KERNELS_SHAPE_H_
+
+#include <array>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+#include "dali/kernels/span.h"
+
+namespace tensor {
+
+
+
+/// @brief Returns the product of all elements in shape
+/// @param shape - shape of a tensor whose elements we count
+template <typename T>
+inline int64_t volume(const T &shape) {
+  int n = shape.size();
+  if (n < 1) {
+    return 0;
+  }
+  int64_t v = shape[0];
+  for (int i = 1; i < n; i++) {
+    v *= shape[i];
+  }
+  return v;
+}
+
+constexpr int DynamicDimensions = -1;
+
+template <int ndim = DynamicDimensions>
+struct TensorShape;
+
+/**
+ * @brief Base class for TensorShape containing common code for iterators and operator[]
+ */
+template <typename Container, int ndim>
+struct TensorShapeBase {
+  using container_type = Container;
+  using value_type = typename container_type::value_type;
+  using size_type = int;
+  using reference = value_type &;
+  using const_reference = const value_type &;
+  using iterator = typename container_type::iterator;
+  using const_iterator = typename container_type::const_iterator;
+  using reverse_iterator = typename container_type::reverse_iterator;
+  using const_reverse_iterator = typename container_type::const_reverse_iterator;
+
+  reference operator[](int d) { return shape[d]; }
+  const_reference operator[](int d) const { return shape[d]; }
+
+  iterator begin() noexcept { return shape.begin(); }
+  iterator end() noexcept { return shape.end(); }
+  const_iterator begin() const noexcept { return shape.begin(); }
+  const_iterator end() const noexcept { return shape.end(); }
+  const_iterator cbegin() const noexcept { return shape.cbegin(); }
+  const_iterator cend() const noexcept { return shape.cend(); }
+  reverse_iterator rbegin() noexcept { return shape.rbegin(); }
+  reverse_iterator rend() noexcept { return shape.rend(); }
+  const_reverse_iterator rbegin() const noexcept { return shape.rbegin(); }
+  const_reverse_iterator rend() const noexcept { return shape.rend(); }
+  const_reverse_iterator crbegin() const noexcept { return shape.crbegin(); }
+  const_reverse_iterator crend() const noexcept { return shape.crend(); }
+
+  size_type size() const { return shape.size(); }
+
+  Container shape;
+  static constexpr int static_ndim = ndim;
+
+  template <int other_ndim>
+  TensorShape<other_ndim> first();
+  template <int other_ndim>
+  TensorShape<other_ndim> last();
+
+  TensorShape<DynamicDimensions> first(int count);
+  TensorShape<DynamicDimensions> last(int count);
+
+ protected:
+  // Disallow instatiation of Base class
+  TensorShapeBase() : shape{} {}
```

Comment:
You can add a comment about `{}` performing 0-fill of `std::array`

Reply:
Done